File: php/recherche.php
<?php
 $dbg = false;
 //if ($dbg) {echo "$_SERVER<br>\n";}
 // Request parameters:
 sanitize_REQUEST_all();
 $dictionary=$_REQUEST['dictionary'];
 $st=$_REQUEST['st'];
 $prst=$_REQUEST['prst'];
 $en=$_REQUEST['en'];
 $pren=$_REQUEST['pren'];
 // Case-sensitive toggle (Wave 2): opt-out from the default case-folded search so HK's
 // phonemic letter case (A=ā, T=ṭ, S=ṣ vs s, ...) isn't conflated for power users. Unset
 // when the form checkbox isn't ticked (browsers omit unchecked checkboxes entirely).
 $case_sensitive = isset($_REQUEST['case_sensitive']) && $_REQUEST['case_sensitive'] == '1';
 $maxhits=$_REQUEST['maxhits'] ;
 // Default + clamp: a missing/empty/0/negative value would otherwise produce
 // "LIMIT 0" (or a SQLite error for negative); a huge value would defeat the
 // form's own cap. Mirror the form's own default (50) and max (1000).
 $maxhits = (int)$maxhits;
 if ($maxhits <= 0) {
  $maxhits = 50;
 } else if ($maxhits > 1000) {
  $maxhits = 1000;
 }
 $parmkeys = array_keys($_REQUEST);
 //$_SERVER['SCRIPT_NAME']='HELLO';
 if ($dbg) {
  foreach($parmkeys as $key) {
   $val = $_REQUEST[$key];
   echo "$key = $val,  "; 
  }
  echo "<br>\n";
 }
// $dictbooks is a simple array. each element an array of 3 strings:
// book number, shortname, fullname.  How to deal with ALL ?
$dictbooks = readbooks();  // simple array.
// get dictname and dictnum from $dictionary
list($dictnum,$dictname) = dictionary_info($dictionary,$dictbooks);
if ($dbg) {
 echo "dictnum=$dictnum, dictname=$dictname<br>\n";
}
$st = trim($st);
$en = trim($en);
if (! ( (strlen($st)>1) || (strlen($en) > 1)) ) {
 fehler("No search has been formulated.");
}
echo "<html><head><title>$dictname: Search Results</title></head>\n",
  "<body bgcolor=\"#ffffff\">\n<h1>$dictname: Search Results</h1>\n";
$where = compute_where($dictnum,$st,$prst,$en,$pren,$case_sensitive);
if ($dbg) {
 echo "<br>where: $where<br>\n";
}
// select statement for sql
$befehl="select id,st,en from tamil where $where order by st collate nocase";
// put in LIMIT

$befehl .= " LIMIT " . (int)$maxhits;  // SQLi guard: $maxhits is $_REQUEST input, already defaulted/clamped above
if ($dbg) {
 echo "befehl: $befehl<br>\n";
}
// get results
$results = selectfromdb($befehl,$case_sensitive);
$nresults = count($results);
if ($dbg) {
 echo "select $nresults results found<br>\n";
}
if ($dictnum == '0') { // user chose ALL dictionaries. Print abbreviations
  echo "<table>\n";
  foreach($dictbooks as $dictbook) {
   list($bnum,$bshort,$blong) = $dictbook;
   // exclude Pahlavi Dictionary, since html does not mention.
   if ($bnum != '4') {
    echo "<tr><td>($bshort) </td><td>=</td><td> $blong</td></tr>\n";
   }
  }
  echo "</table><br>\n";
}
echo "<table cellspacing=3>";
for($i=0;$i<$nresults;$i++) {
 $result = $results[$i];
 $hitnr = $i + 1;
 list($id,$st,$en) = $result;  // NOTE reused variables
 if ($dictnum == '0') {
  $idx = intval($id) - 1;
  list($bnum,$bshort,$blong) = $dictbooks[$idx];
  $buchaus = "<td valign='top'>($bshort)</td>";
 } else {
  $buchaus = "";
 }
 //
 echo "<tr><td align=right valign=top>$hitnr</td>\n";
 echo "$buchaus";
 echo "<td valign=top> <b>$st</b></td>\n";
 // otherwise, get unprintable characters.
 // Note: better would be to use 'iconv' on ganz.txt, then remake tamil.sqlite.
 // Perl code seems to do this conversion on its own.
 $en1 = iconv("Windows-1252","UTF-8",$en); 
 echo "<td valign=top>$en1</td></tr>\n";
}
echo "</table>\n";
if ($nresults == 0) {
 echo "No entries found.<br>";
}
echo "</body></html>";

/************************************************************
functions
*/
function fehler($msg) {
 // Escape: $msg may embed user input (e.g. the unknown dictionary code),
 // which would otherwise be reflected XSS. All fehler messages are plain text.
 $ans = "<h4>" . htmlspecialchars($msg, ENT_QUOTES) . "</h4>\n</body></html>\n";
 echo $ans;
 exit;
}
function readbooks() {
 $filename = '../dat/books'; // a text file
 if (! file_exists($filename)) {
  fehler("missing file $filename");
 }
 $lines = file($filename,FILE_IGNORE_NEW_LINES);
 $dbg = false;
 $ans = array();
 foreach($lines as $line) {
  if ($dbg) {echo "$filename: $line<br>";}
  if (! preg_match('/^([1-4]) (.*?) (.*)$/',$line,$matches)) {
   fehler("readbooks error: $line");
  }
  // book number, shortname, fullname
  $ans1 = array($matches[1],$matches[2],$matches[3]);
  $ans[] = $ans1;
 }
 return $ans;
}
function dictionary_info($dictionary,$dictbooks) {
 if ($dictionary == 'all') {
  $dictnum = '0';
  $dictname = 'All dictionaries';
 } else {
  $found = false;
  for($i=0; $i < count($dictbooks); $i++) {
   $dictbook = $dictbooks[$i];
   list($dictnumstr,$dictshort,$dictlong) = $dictbook;
   if ($dictionary == $dictshort) {
    $dictnum = $dictnumstr;
    $dictname = $dictlong;
    $found = true;
    break;
   }
  }
  if (! $found) {
   fehler("No dictonary has been selected.$dictionary");  // exits
  }
 }
 return array($dictnum,$dictname);
}
function compute_where($dictnum,$st,$prst,$en,$pren,$case_sensitive=false) {
 // construct the 'where' string for sql search
 // sqlite file has 3 fields id (= dictnum), st (headword), en (text)
 // ---- id dictnum
 // assume $dictnum is a string
 $dbg = false;
 if ($dictnum == '0') {
  // 'all', exclude the 4th (Pahlavi dictionary)
  $where = "id<4";
 } else {
  $where = "id=$dictnum";
 }
 // ---- st
 if ($st != "") {
  $temp = where1($st,'st',$prst,$case_sensitive);
  $where .= " and $temp";
 }
 if ($en != "") {
  $temp = where1($en,'en',$pren,$case_sensitive);
  $where .= " and $temp";
 }
 return $where;
}
function where1($var,$varname,$pr,$case_sensitive=false) {
 $wb = "\\b"; // word begin in regexp
 $we = "\\b"; // word end
 // allow $var to have multiple words, separated by one or more spaces
 $var = trim($var);
 $parts = preg_split('/ +/',$var);
 $ans = "";
 // Case-sensitive mode (Wave 2) skips the lower()/strtolower() folding below. Two things
 // beyond that are needed for it to actually take effect (both handled here / in
 // selectfromdb()): SQLite's REGEXP infix operator is hardwired to a function literally
 // named 'regexp', so a case-sensitive variant must be called via ordinary function-call
 // syntax (regexp_cs(pattern, col)), not as an infix operator -- verified empirically,
 // "col regexp_cs 'pattern'" is a SQL syntax error. And SQLite's LIKE is
 // case-INsensitive for ASCII by default regardless of lower()-folding in the query, so
 // selectfromdb() sets PRAGMA case_sensitive_like when $case_sensitive is true.
 // in sqlite select, lowdata puts the result text into lower case for regexp (unless case-sensitive)
 $lowdata = $case_sensitive ? $varname : "lower($varname)";  //lower is sqlite function name
 for($ipart=0;$ipart < count($parts); $ipart++) {
  $part = $parts[$ipart];
  $part_term = $case_sensitive ? $part : strtolower($part);
  // LIKE-branch escaping: neutralize the LIKE metacharacters '%' and '_' (and
  // '\' itself, the escape char) so a literal '%'/'_' in the query matches
  // itself instead of acting as a wildcard -- see the ESCAPE clause below.
  // Then escape ' for the SQLite string literal (SQLi guard).
  $x = str_replace(array('\\', '%', '_'), array('\\\\', '\\%', '\\_'), $part_term);
  $x = str_replace("'", "''", $x);
  // The regexp branches feed the value into _sqliteRegexp()/_sqliteRegexpCS() -> preg_match()
  // as a PCRE pattern, so the term must ALSO be preg_quote()d -- otherwise regex
  // metacharacters inject and a term like "(a+)+" is a catastrophic-backtracking ReDoS
  // (run per row). $wb/$we are the intended \b word-boundary anchors and are left active.
  $xr = str_replace("'", "''", preg_quote($part_term, '/'));
  if ($pr == "exact") {
    $pattern = "$wb$xr$we";
  } else if ($pr == "prefix") {
    $pattern = "$wb$xr";
  } else if ($pr == "suffix") {
    $pattern = "$xr$we";
  } else {
    $pattern = null; // substring: LIKE, not regexp
  }
  if ($pattern !== null) {
    $ans1 = $case_sensitive
      ? "(regexp_cs('$pattern', $lowdata))"
      : "($lowdata regexp '$pattern')";
  } else { // substring
    $ans1 ="($lowdata like '%$x%' ESCAPE '\\')";
  }
  if ($ipart != 0) {
   $ans .= " and $ans1";
  }else {
   $ans .= $ans1;
  }
  // echo "where1 dbg: ipart=$ipart, part=$part, ans1=$ans1, ans=$ans<br>\n";
 }
 return $ans;
}
function sanitize_REQUEST_all() { 
 $parmkeys = array_keys($_REQUEST);
 foreach($parmkeys as $key) {
  $old = $_REQUEST[$key];
  // remove all HTML tags from a string
  // another could be FILTER_SANITIZE_URL
  //$new = filter_var($old,FILTER_SANITIZE_STRING); //deprecated php 8.2
  // $new = filter_var($old,FILTER_SANITIZE_URL); // dog cat not found
  $new = filter_var($old,FILTER_UNSAFE_RAW);
  $_REQUEST[$key] = $new;
 }
}
function selectfromdb($sql,$case_sensitive=false) {
 $sqlitefile = "../sqlite/tamil.sqlite";
 try {
   $file_db = new PDO('sqlite:' .$sqlitefile);
   $file_db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
   $status=true;
  } catch (PDOException $e) {
   $file_db = null;
   fehler("Cannot open " . $sqlitefile . "\n");
  }
 $file_db->sqliteCreateFunction('regexp', '_sqliteRegexp', 2);
 $file_db->sqliteCreateFunction('regexp_cs', '_sqliteRegexpCS', 2);
 if ($case_sensitive) {
  // SQLite's LIKE is case-insensitive for ASCII by default, independent of any
  // lower()/strtolower() folding done in the query itself -- this pragma is what
  // actually makes the 'substring' (LIKE) match mode case-sensitive. Connection-local,
  // so it never affects any other request.
  $file_db->exec('PRAGMA case_sensitive_like = ON;');
 }
 //echo "sql=$sql<br>\n";
 try {
  $result = $file_db->query($sql);
 } catch (PDOException $e)  {
  fehler("selectfromdb error: $e");
 }
 $ansarr = array();
 foreach($result as $m) {
  $rec = array($m['id'],$m['st'],$m['en']);
  $ansarr[] = $rec;
 }
 // close db
 $file_db = null;
 return $ansarr;
}
//$nreg=0;
function _sqliteRegexp($pattern, $string) {
 /*
  GLOBAL $nreg;
  $nreg = $nreg + 1;
  if ($nreg < 10) {echo "_sqliteRegexp $nreg $string<br>\n";}
 */
    #if(preg_match('/^'.$pattern.'$/i', $string)) {
    if(preg_match('/'.$pattern.'/i', $string)) {
        return true;
    }
    return false;
}
function _sqliteRegexpCS($pattern, $string) {
    // Case-sensitive counterpart of _sqliteRegexp() -- same pattern, no /i flag.
    if(preg_match('/'.$pattern.'/', $string)) {
        return true;
    }
    return false;
}

?>

File: perl/recherche.pl
#!"C:\xampp\perl\bin\perl.exe"
#!/usr/bin/perl
#!/vol/perl/bin/perl
# 11.2.2003 SW: Recherche in der tamil-Datenbank 
# 12.8.2008 EJF: modified for sanskrit-lexicon website configuration
# 12.9.2008 EJF: omitted cpd (Concise Pahlavi Dictionary) at request of T. Malten
# 05.31.2015 EJF: Modify to work with sqlite3.
#   Testing was done with XAMPP server under Windows Vista.
print "Content-type: text/html\n\n";
# do 'cgi-include2.pl';
# BEGIN cgi-include2.pl
#!"C:\xampp\perl\bin\perl.exe"
$data1=$ARGV[0] || $ENV{'QUERY_STRING'} || '';
$inq=$ENV{'CONTENT_LENGTH'};
if ($inq) {
  read (STDIN,$data,$inq);
  $data="$data1&$data" if $data1}
else {$data=$data1}
$tr='^?'; $trEsc='\^\?';

$content_type=$ENV{'CONTENT_TYPE'};
$content_type=~/multipart\/form-data\; boundary\=(.+)$/;
#print "=i=content_type==$content_type==<br>\n",
#  "=1=$1==<br>\n";

if ((defined (my $content_type=$ENV{'CONTENT_TYPE'})) &&
  ($content_type=~/multipart\/form-data\; boundary\=(.+)$/)) {
  &zerleg2($data)}
else {&zerleg($data)}
$me=$ENV{'SCRIPT_NAME'};
#print "data=$data.\n";
#$,=","; @F=%F; @n=%n;
#print "data=$data.\nme=$me.\na=@a.\nb=@b.\naorig=@aorig.\n",
#  "borig=@borig.\nn=@n.\nF=@F.\n";
#for (sort keys %F) {print "F{$_}=$F{$_}<br>\n"}

#-----------
sub zerleg {
#-----------
my $data=shift;
my ($a,$b,$n);
$n=0; @a=(); @b=(); @aorig=(); @borig=(); %n=(); %F=();
for (split(/&/,$data)) {
  ($a,$b)=split(/=/,$_);
  $aorig[$n]=$a; $borig[$n]=$b;
  $a=~tr/+/ /;
  $a=~s/%([a-fA-F0-9][a-fA-F0-9])/pack("C", hex($1))/eg;
  $b=~tr/+/ /;
  $b=~s/%([a-fA-F0-9][a-fA-F0-9])/pack("C", hex($1))/eg;
  $b=~s/^[\s\n]+//; $b=~s/[\s\n]+$//;
  eval $weitereErsetzung if $weitereErsetzung;
  $a[$n]=$a; $b[$n]=$b; $n{$a}=$n++;
  if (defined $F{$a}) {$F{$a}.="$tr$b"} else {$F{$a}=$b}}}

#-----------
sub zerleg2 {
#-----------
my $data=shift;
my ($boundary,@list,$part,$a,$b,$n);
$boundary='--'.$1; # RFC1967
@list=split(/$boundary/,$data);
#print "Boundary=$boundary,list=\n",join("\n===",@list);
for $part (@list) {
  ($head,$b)=(split(/\r\n\r\n|\n\n/,$part,2));
  next unless $head;
  next unless $b;
  $b=~s/(\r\n|\n)$//;
  $head=~/name="([^"]*)"/;
  $a=$1;
  #if ($a eq "w") {print "[$part]\n"}
  $head=~/filename="([^"]*)"/;
  $F{"$a.fname"}=$1 if $1;
  #$b=~s/\r?\n$//;
  #print "a=$a, b=$b<br>\n";
  #$b=~s/^[\n\s]+//; $b=~s/[\n\s]+$//;
  #$b=~/\r\s/\s/g; $b=~/\r/\s/g; $b=~s/\s+\n/\n$/g;
  #$b=~s/[\r\s]+$//;
  eval $weitereErsetzung if $weitereErsetzung;
  $a[$n]=$a; $b[$n]=$b; $n{$a}=$n++;
  if (defined $F{$a}) {$F{$a}.="$tr$b"} else {$F{$a}=$b}}}

# END cgi-include2.pl
$dictionary=$F{"dictionary"};
$st=$F{'st'}; $prst=$F{'prst'};
$en=$F{'en'}; $pren=$F{'pren'};
$maxhits=$F{'maxhits'} || 100;
#print "st=$st,prst=$prst,en=$en,pren=$pren<br>\n";
$dir="../dat"; 
#%dict=(
#  'mwd' => "Cologne Digital Sanskrit Lexicon",
#  'cap' => "Capeller's Sanskrit-English Dictionary",
#  'otl' => "Cologne Online Tamil Lexicon",
#  'cpd' => "Concise Pahlavi Dictionary");
#$dictname=$dict{$dictionary} || &fehler (
#  "No dictonary has been selected.");

#%dict=('all'=>0, 'mwd'=>1, 'cap'=>2, 'otl'=>3, 'cpd'=>4);
#@dictshort=('all','mwd','cap','otl','cpd');
#@dict=("All dictionaries","Cologne Digital Sanskrit Lexicon",
#  "Capeller's Sanskrit-English Dictionary",
#  "Cologne Online Tamil Lexicon","Concise Pahlavi Dictionary");
&readbooks;
if ($dictionary eq 'all') {$dictnum=0}
else {defined ($dictnum=$dict{$dictionary}) || &fehler (
  "No dictonary has been selected.$dictionary")}
$dictname=$dict[$dictnum];

print
  "<html><head><title>$dictname: Search Results</title></head>\n",
  "<body bgcolor=\"#ffffff\">\n<h1>$dictname: Search Results</h1>\n";
#unless ($st || $en) {&fehler("No search has been formulated.")}
unless (($st && (length($st)>1)) || ($en && (length($en)>1))) {
  &fehler("No search has been formulated.")}
# $befehl="select id,st,en from $dictionary where ";
#EJF changed 'buch' to 'id' to agree with 'tamil' sql table
# EJF $where=$dictnum ? "id=$dictnum" : "";
$where=$dictnum ? "id=$dictnum" : "id<4";  # exclude cpd with id=4.
&where($st,'st',$prst);
&where($en,'en',$pren);
unless ($where) {&fehler ("No useable search has been formulated.")}
#$befehl="select id,st,en from tamil where $where order by st";
# 'collate nocase' needed in sqlite3, otherwise ordering is case sensitive.
$befehl="select id,st,en from tamil where $where order by st collate nocase";
#print "SQL query: $befehl \n <br>";
#print "data=$data<br>\nmaxhits=$maxhits";
#&opendb("webapps");
$dbh = opendb();
#print "back from opendb\n<br/>";
#$befehl = "select * from tamil where (st regexp 'rA') ";
#print "SQL query: $befehl \n <br>";
$sth=$dbh->prepare($befehl);
($sth->execute) || &fehler("Error $!\n".$sth->errstr);
#print "dictnum = $dictnum\n<br/>";
if (!$dictnum) {
  print "<table>\n";
  for ($i=1;$i<=$#dict;++$i) {print
    "<tr><td>($dictshort[$i]) </td><td>=</td><td> $dict[$i]</td></tr>\n"}
  print "</table><br>\n"}
print "<table cellspacing=3>";
$hitnr=0;
while (($id,$st,$en)=$sth->fetchrow_array) {
  last if (++$hitnr>$maxhits);
  $st=~s/^\. +//; $st=~s/ +\.$//; $en=~s/^\. +//; $en=~s/ +\.$//;
  $buchaus=$dictnum ? "" : "<td valign=top>($dictshort[$id])</td>";
  print "<tr><td align=right valign=top>$hitnr</td>\n",
    "$buchaus<td valign=top> <b>$st</b></td>\n",
    "<td valign=top>$en</td></tr>\n"}
print "</table>\n";
print "No entries found." unless $hitnr;
#print "hitnr = $hitnr\n<br/>";
print "</body></html>";
&closedb;

#------------------------
sub where {
# Error POSIX class [:<:] unknown in regex; marked by <-- HERE in m/[[:<:] <-- HERE ]deva[[:>:]]/ 
#------------------------
my ($var,$varname,$pr)=@_;
return unless $var;
$wb='[[:<:]]';   # word begin in regexp
$we='[[:>:]]';   # word end
$wb = "\\b"; # May 31, 2015.  For sqlite under XAMPP 
$we = "\\b";
for (split(/ +/,$var)) {
  next if $_ eq '';
  s/\\/\\\\/g; s/\'/\\\'/g; s/\%/\\\%/g;
  $where.=" and " if $where;
  #if ($pr eq "exact") {$where.="($varname like '% $_ %')"}
  #elsif ($pr eq "prefix") {$where.="($varname like '% $_%')"}
  $regexp = 'regexp';  # EJF prior code
  #print "check dollar_ = $_<br>";
  my $x = lc($_);
  if ($pr eq "exact") {
    #$where.="($varname $regexp '$wb$_$we')"}
    #$where.="(lower($varname) $regexp '$wb$_$we')"}   # 12-07-2022
    $where.="(lower($varname) $regexp '$wb$x$we')"}   # 12-07-2022
  elsif ($pr eq "prefix") {
    #$where.="(lower($varname) $regexp '$wb$_')"}  #12-07-2022
    $where.="(lower($varname) $regexp '$wb$x')"}  #12-07-2022
  else {
    #$where.="(lower($varname) like '%$_%')"}}} #12-07-2022
    $where.="(lower($varname) like '%$x%')"}}} #12-07-2022

#------------------------
sub fehler {print "<h4>$_[0]</h4>\n</body></html>\n"; exit}
#------------------------
sub opendb { # EJF May 31, 2015. for 
# ignore dbname passed as argument
require DBI;
my  $dbh = DBI->connect ("DBI:SQLite:dbname=../sqlite/tamil.sqlite","","")
             || die "Could not connect to database: "
             . DBI-> errstr;
#print "dbh = $dbh\n";
    return $dbh;

}

#------------------------
sub closedb {
#------------------------
$sth->finish();
$dbh->disconnect()}

#------------
sub readbooks {
#------------
open (F,"<$dir/books") or &fehler ("no open (F,\"&lt;$dir/books\"):$!");
read (F,$books,10000);
close F;
# example line: 1 mwd Cologne Digital Sanskrit Lexicon
%dict=(0); @dict=(); @dictshort=();
for (split(/\n/,$books)) {
  s/^[\s\n]+//; s/[\s\n]+$//; 
    #print "booksline=$_<br>";
  if (/(\d+)\s+(\S+)\s+(.+)$/) {
    #print "1=$1,2=$2,3=$3,<br>\n";
    $dict{$2}=$1; $dictshort[$1]=$2; $dict[$1]=$3}}
#print "<br/>readbooks: dict=$#dict<br>\n";
}
#=======================================================================

